ImageInLib/ImageInLib/src/imageInterpolation.c

```diff
@@ -3,257 +3,145 @@
 #include "imageInterpolation.h"
 #include<math.h>
 
-bool nearestNeighborInterpolation(dataType** originalImage, dataType** newImage, size_t imageLength, size_t imageWidth, size_t imageHeight, dataType originalSpacing, dataType newSpacing)
-{
-    if (originalImage == NULL || newImage == NULL)
-        return false;
+//3D images
 
-    size_t i, j, k, kn, x;
-    dataType k_int, k1, k2;
-
-    k_int = 0; kn = 0;
-
-    for (k = 0; k < imageHeight - 1; k++) {
-        k1 = originalSpacing * k;
-        k2 = originalSpacing * (k + 1);
-        do {
-            for (i = 0; i < imageLength; i++) {
-                for (j = 0; j < imageWidth; j++) {
-                    x = x_new(i, j, imageLength);
-                    if (k_int == 0) {
-                        newImage[kn][x] = originalImage[k][x];
-                    }
-                    else {
-                        if ((k_int - k1) < (k2 - k_int)) {
-                            newImage[kn][x] = originalImage[k][x];
-                        }
-                        else {
-                            newImage[kn][x] = originalImage[k + 1][x];
-                        }
-                    }
-                }
-            }
-            k_int = k_int + newSpacing;
-            kn = kn + 1;
-        } while (k_int < k2);
-    }
+Point3D getImageCoordFromRealCoord3D(Point3D image_point, Point3D image_origin, VoxelSpacing image_spacing, OrientationMatrix orientation) {
+    Point3D resultPoint;
+    resultPoint.x = image_origin.x + image_spacing.sx * orientation.v1.x * image_point.x;
+    resultPoint.y = image_origin.y + image_spacing.sy * orientation.v2.y * image_point.y;
+    resultPoint.z = image_origin.z + image_spacing.sz * orientation.v3.z * image_point.z;
+    return resultPoint;
+}
 
-    return true;
+Point3D getRealCoordFomImageCoord3D(Point3D real_point, Point3D real_origin, VoxelSpacing real_spacing, OrientationMatrix orientation) {
+    Point3D resultPoint;
+    resultPoint.x = (real_point.x - real_origin.x) / real_spacing.sx * orientation.v1.x;
```

Comment:
denominators are not correct in my opinion
it should be like this

resultPoint.x = (real_point.x - real_origin.x) / (real_spacing.sx * orientation.v1.x);

Reply:
yes definitively